### ISSUE TITLE
[upstream/Apple] Kokoro ANE crashes on M5/macOS 26.5 (libBNNS iSTFT + GPURNNOps) — worked around via ane-tail-gpu; tracking Apple fix

### ISSUE THREAD

**Issue by Alex-Wengg** [FluidInference/FluidAudio#667]
## Summary
Kokoro ANE (both `english` and `mandarin`) **crashes on the 2nd+ CoreML prediction within a single process** on **M5 Pro / macOS 26.5 (25F71)**. A single synthesis works every time and produces correct audio; the crash is triggered by *repeated* synthesis on a reused manager — independent of phrase length and of compute-unit routing.

## Two crash modes (depend on compute-unit routing)
- `--compute-units default` (prosody/noise/tail = `.all`): GPU path asserts
  ```
  MetalPerformanceShadersGraph .../GPURNNOps.mm:1565: failed assertion `JIT not supported'
  ```
  (SIGABRT / exit 134) on the 2nd prediction — the prosody RNN.
- `--compute-units all-ane` or `cpu-only`: CPU path segfaults
  ```
  EXC_BAD_ACCESS (SIGSEGV) in libBNNS.dylib
    BNNSGraphContextExecute_v2
    E5RT::Ops::BnnsCpuInferenceOperation::ExecuteSync()
  ```
  (exit 139). **Nondeterministic** — sometimes phrase 1 completes, sometimes 0 completions.

## Minimal repro (no benchmark corpus needed)
```
# single synthesis — WORKS, every time, both variants:
swift run fluidaudiocli tts "If you want something, go get it." \
  --backend kokoro-ane --variant english --output ok.wav            # exit 0
swift run fluidaudiocli tts "我只想和你一起，看遍这世界的美好。" \
  --backend kokoro-ane --variant mandarin --voice zf_001 --output ok_zh.wav  # exit 0

# repeated synthesis (same short phrase x5) — CRASHES on the 2nd prediction:
printf 'If you want something, go get it.\n%.0s' {1..5} > rep.txt
swift run fluidaudiocli tts-benchmark --backend kokoro-ane --variant english \
  --corpus-path rep.txt --skip-asr                                  # exit 134 (default) / 139 (all-ane,cpu-only)
```
The exact same phrase that succeeds standalone crashes as the 2nd synthesis in the loop.

## Diagnosis
- Not phrase-length: the crashing phrase synthesizes fine standalone.
- Not compute-units: reproduces under `default`, `all-ane`, and `cpu-only` (GPU assert vs CPU BNNS segfault).
- Not ASR contention: reproduces with `--skip-asr`.
- It is the **2nd+ prediction on the reused MLModel graphs**; nondeterminism points to a driver/JIT state bug. Both failing frames are inside Apple frameworks (MPSGraph / libBNNS) — not FluidAudio code.

## Impact
Kokoro is the only Mandarin backend. It works for one-off / low-rate synthesis on M5 but is not safe for sustained use (a TTS service synthesizing many utterances in one process will crash). `Documentation/TTS/Benchmarks.md` carries the Kokoro rows from M2.

## Workarounds attempted (all fail)
`--compute-units all-ane` and `cpu-only` (moves the RNN off GPU) — still segfaults in BNNS. No routing avoids it.

## Suggested next steps
- File an Apple Feedback with the minimal repro (repeated CoreML prediction on the Kokoro prosody RNN / vocoder graphs, M5 + macOS 26.5).
- Investigate re-exporting the affected graphs with fixed (bucketed) shapes so they don't re-plan/re-JIT per prediction — mirrors the Supertonic L-bucketing approach.

---

## Fix attempts (2026-06-07)

**1. Re-export with fixed/bucketed shapes — infeasible here.** The Kokoro CoreML graphs are third-party (`laishere/kokoro-coreml`, re-hosted as `FluidInference/kokoro-82m-coreml`); there is no conversion source in `mobius`, so baking fixed shapes into the models would mean rebuilding laishere's 7-graph pipeline from the PyTorch model.

**2. Runtime fixed-shape padding in Swift — pipeline runs, but the tail iSTFT still segfaults BNNS.** Implemented a prototype that pads every utterance to a single `(tokens, frames)` shape (token padding via `attention_mask=0`; frame padding via trailing EOS-token durations; output trimmed back). Findings:
- Under `all-ane`, a *single repeated shape* is stable, but the **2nd distinct shape** crashes — so a single fixed shape per process is required (multi-bucket does not help: the 2nd bucket is a 2nd distinct shape).
- With a single fixed shape, the 6 upstream stages (albert→vocoder) run fine, but the **tail (iSTFT) stage segfaults in `libBNNS` on the padded tensor** (`x_pre [1,128,61441]` at 512 frames; `vocoder DONE` prints, `tail DONE` never does). Shrinking the fixed frame target (1024→512) did not avoid it.

Net: neither path resolves it from the FluidAudio side — the crash is inside Apple's BNNS on these graphs on M5/macOS 26.5. Prototype reverted (not shipped).

**Untried lever:** route only the tail (iSTFT) stage to a non-BNNS compute unit (e.g. `.cpuAndGPU`) while keeping the RNN-bearing prosody stage off GPU — may dodge both the BNNS tail crash and the GPU `GPURNNOps` crash. Speculative.

---

## Workaround found & shipped (2026-06-07) — `ane-tail-gpu` preset

**Root-caused to two stages on two compute paths:** the prosody **RNN** crashes on the **GPU** (`GPURNNOps` JIT assert), and the tail **iSTFT** crashes on **CPU/BNNS** (`libBNNS` SIGSEGV). The stock presets always put at least one of them on its bad path.

**Fix:** a new `TtsComputeUnitPreset.aneTailGpu` (`--compute-units ane-tail-gpu`) routes every stage to `.cpuAndNeuralEngine` **except the tail (iSTFT) → `.cpuAndGPU`** — RNN stays off the GPU, iSTFT stays off BNNS. Dynamic shapes, no padding.

**Verified on M5 Pro / macOS 26.5** (full 100-phrase minimax runs, clean):
- en: agg RTFx **25.4×**, synth 314/410 ms, WER **10.33%** / CER 3.72% (vs M2 7.47× / 10.8% / 4.0%)
- zh: agg RTFx **20.1×**, synth 201/286 ms, CER **3.81%** via whisper-large-v3-turbo (vs M2 4.01%)

Quality matches M2, confirming the routing change is numerically faithful. Shipped in commit `1d1ed18`; Benchmarks.md updated.

The underlying Apple framework bug (stock `default`/`all-ane`/`cpu-only` crash) remains — leaving this issue open to track an Apple Feedback / eventual OS fix, but Kokoro is now usable on M5 via the preset.

---

## Isolated root cause (2026-06-07) — two distinct Apple bugs, single-graph repros

Reduced each crash to **one Kokoro CoreML graph + coremltools** (no FluidAudio, no pipeline). Both are data-independent (same random input that passes on one backend crashes on the other).

| Stage | on `CPU_AND_GPU` | on `CPU_AND_NE` / `CPU_ONLY` | crashing op |
|---|---|---|---|
| **KokoroProsody** (RNN) | ❌ `GPURNNOps.mm:1565 'JIT not supported'` (2nd predict) | ✅ survives (varied shapes) | MPSGraph GPU RNN |
| **KokoroTail** (iSTFT) | ✅ survives (varied shapes) | ❌ `EXC_BAD_ACCESS` in `libBNNS` (1st predict) | BNNS FFT/iSTFT |

They fail on **opposite** backends — which is exactly why the `ane-tail-gpu` preset (every stage on ANE/CPU *except* the tail iSTFT on GPU) is the unique stable routing.

### Minimal repro A — prosody RNN, GPU JIT assert
```python
import numpy as np, coremltools as ct
m = ct.models.CompiledMLModel("KokoroProsody.mlmodelc", ct.ComputeUnit.CPU_AND_GPU)
f = lambda N: {"en": np.random.randn(1,640,N).astype(np.float16),
               "style_s": np.random.randn(1,128).astype(np.float16)}
m.predict(f(200))   # #1 OK
m.predict(f(200))   # #2 -> GPURNNOps.mm:1565 failed assertion 'JIT not supported'
```

### Minimal repro B — tail iSTFT, BNNS segfault
```python
import numpy as np, coremltools as ct
m = ct.models.CompiledMLModel("KokoroTail.mlmodelc", ct.ComputeUnit.CPU_AND_NE)  # or CPU_ONLY
m.predict({"x_pre": (np.random.randn(1,128,24001)*0.1).astype(np.float32)})  # -> SIGSEGV in libBNNS
# identical input runs fine under ct.ComputeUnit.CPU_AND_GPU
```

Host: Apple M5 Pro, macOS 26.5 (25F71), coremltools 9.0. Both ready to attach to an Apple Feedback. The `ane-tail-gpu` preset is the in-app mitigation until Apple fixes the GPU RNN JIT and the BNNS FFT paths.
